React to latest Roslyn nullability changes

GrabYourPitchforks opened the pull request:
Follow-up to https://github.com/dotnet/runtime/pull/35782, which addressed the corelib side of things. This PR addresses the libraries side.

Due to https://github.com/dotnet/runtime/issues/34417#issuecomment-624428649 I can't remove the override in __Versions.props__ just yet. But the latest nightly compiler toolchain fixes that blocking issue, so this PR updates the explicit override to point to the latest nightly to demonstrate that we do compile cleanly against it.

I plan on reverting my changes to __Versions.props__ as part of checkin, assuming the updated code also compiles cleanly on the older toolset. If the updated code doesn't compile cleanly on the older toolset, then the best course of action might be to wait until a new drop of Arcade comes out which picks up the updated compiler, then update Arcade, merge the contents of this PR, _and_ remove the override from __Versions.props__ all in one large PR.

Contributes to https://github.com/dotnet/runtime/issues/34417.

GrabYourPitchforks committed:
React to latest Roslyn nullability changes

GrabYourPitchforks commented on eng/Versions.props:
```diff
@@ -19,7 +19,8 @@
     <UsingToolXliff>false</UsingToolXliff>
     <!-- Downgrade compiler version to fix nullability issues: https://github.com/dotnet/runtime/issues/34417. -->
-    <MicrosoftNetCompilersToolsetVersion>3.6.0-2.20166.2</MicrosoftNetCompilersToolsetVersion>
+    <!-- <MicrosoftNetCompilersToolsetVersion>3.6.0-2.20166.2</MicrosoftNetCompilersToolsetVersion> -->
```
Reminder: I plan on reverting the changes to this file if possible. It's only here to demonstrate which toolsets we're using.

stephentoub replied:
Any reason we can't keep the newest version of the compiler?  I'm unclear why we haven't taken a newer one recently.

GrabYourPitchforks replied:
Our props file explicitly pinned to an older version partially due to breaking changes in how nullability annotations are processed. It also turns out there's a bug in the compiler toolset version currently carried by Arcade (see https://github.com/dotnet/roslyn/issues/44000) that causes build failures. That bug was addressed in the latest nightly compiler toolchain, which is why for this PR I've leapfrogged the toolset version that comes with Arcade and instead pinned the latest fixed version.

stephentoub replied:
> I've leapfrogged the toolset version that comes with Arcade and instead pinned the latest fixed version.

Ok, thanks.  I didn't understand the "I plan on reverting the changes to this file" part.  I thought you were saying you were making the changes to the annotations but wouldn't actually check in the switch to the newer compiler.  Since we were already pinning a version, I suggest you keep your change to pin the latest version that works.  If that's what you were already planning on doing, great :)

GrabYourPitchforks replied:
I \[not-so\] secretly just wanted the nint/nuint support that the latest toolchain brings. ;)

GrabYourPitchforks commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:
```diff
@@ -755,7 +755,16 @@ private bool TryTakeWithNoTimeValidation([MaybeNullWhen(false)] out T item, int
+
+            if (waitForSemaphoreWasSuccessful)
+            {
+                Debug.Assert(item != null);
```
I'm not sure if this assertion is actually correct. Is it possible to have - say - `BlockingCollection<string?>`? If that's legal, we should remove this assert.

stephentoub replied:
> Is it possible to have - say - `BlockingCollection<string?>`?

Yes, that's valid.  Please revert the added code.  It's a little disheartening this didn't cause any existing tests to fail.  Sigh.

GrabYourPitchforks replied:
I'll remove the assertion and add an explicit unit test.

GrabYourPitchforks commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -354,7 +354,7 @@ public abstract partial class Array : System.Collections.ICollection, System.Col
         public static int LastIndexOf<T>(T[] array, T value, int startIndex) { throw null; }
         public static int LastIndexOf<T>(T[] array, T value, int startIndex, int count) { throw null; }
-        public static void Resize<T>([System.Diagnostics.CodeAnalysis.NotNullAttribute] ref T[]? array, int newSize) { }
+        public static void Resize<T>([System.Diagnostics.CodeAnalysis.NotNullAttribute] ref T[]? array, int newSize) { throw null; }
```
Interesting consequence of the recent `[NotNull]` change is that the compiler now treats it as an _out_-like parameter: you must assign a non-null value to it before method exit. So this means our ref assemblies' dummy method bodies need to throw, just as they would for any other method which has an _out_ parameter.

I didn't see any other cases where this needed to be changed, but might have implications for our "generate ref asms automatically" tool.

GrabYourPitchforks commented:
I forgot to sweep up the _System.Text.Json_ files in this PR. Derp. Will get to it later.

GrabYourPitchforks commented:
@steveharter @layomia @Jozkee Can I get your eyes specifically on the commit https://github.com/dotnet/runtime/pull/36104/commits/29a837b8d436dc5925d977a92a77c9de6070b59e in this PR? There was a bit to untangle and I was trying to follow other established patterns within System.Text.Json.

stephentoub commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:
```diff
@@ -755,7 +755,16 @@ private bool TryTakeWithNoTimeValidation([MaybeNullWhen(false)] out T item, int
+#pragma warning disable CS8762
+            // Compiler can't automatically deduce 'item' has a non-null value when returning false.
             return waitForSemaphoreWasSuccessful;
+#pragma warning restore CS8762
```
There's no way to suppress this warning with a well-placed `!` somewhere?  The clutter caused by `!` is bad enough, but adding a bunch of pragmas to suppress false positive nullable warnings just makes me sad.
cc: @jcouv

jcouv replied:
It's still not great, but possibly a bit better: add a `return true;` below the `Debug.Assert(item != null);` at line 761.

From the language/compiler perspective, it may be useful to allow `return waitForSemaphoreWasSuccessful!;` to solve this.

---
In reply to: [422141700](https://github.com/dotnet/runtime/pull/36104#discussion_r422141700) [](ancestors = 422141700)

stephentoub replied:
> From the language/compiler perspective, it may be useful to allow return waitForSemaphoreWasSuccessful!; to solve this.

@jcouv, yeah, let's do that :smile:

GrabYourPitchforks replied:
I did consider using the `if (condition) { assert(); return true; } else { return false; }` pattern, but it produces different IL and codegen than a simple `return condition;`. I tried to avoid making changes that affected the codegen in any way.

jcouv replied:
Argh, I missed that the `Debug.Assert(item != null)` doesn't work here. An alternative would be `_ = item!;` to pretend that it's not null, but that smells in this case, and also that's not yet implemented.
I'll push on allowing suppression on `return someBoolValue!;`.

jcouv replied:
Another option (also smells) is to suppress on places that assign to `item` (pretend like we're never assigning maybe-null values)

---
In reply to: [422242826](https://github.com/dotnet/runtime/pull/36104#discussion_r422242826) [](ancestors = 422242826)

GrabYourPitchforks replied:
Yeah, using the ! operator inappropriately early in the method strikes me as undesirable. I'll leave the pragma for now and add a code comment directing to https://github.com/dotnet/runtime/issues/36132. Once the language team settles on guidance for this we can react to it here.

Much appreciate your responsiveness on this! :)

stephentoub commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs:
```diff
@@ -755,7 +755,16 @@ private bool TryTakeWithNoTimeValidation([MaybeNullWhen(false)] out T item, int
+            }
+
+#pragma warning disable CS8762
+            // Compiler can't automatically deduce 'item' has a non-null value when returning false.
```
I don't understand the comment.  `item` is allowed to be null if the method returns false.

stephentoub commented on src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs:
```diff
@@ -219,7 +219,11 @@ private bool TrySteal(out T result, bool take)
+                    // Compiler can't automatically deduce that 'result' is set to a valid value
+                    // (which may be null if T allows it) at this exit point.
                     return true;
+#pragma warning restore CS8762
```
Same question

stephentoub commented on src/libraries/System.Net.Mail/src/System/Net/Mime/HeaderCollection.cs:
```diff
@@ -22,9 +22,9 @@ internal HeaderCollection() : base(StringComparer.OrdinalIgnoreCase)
+#pragma warning disable CS8765 // Nullability of parameter 'name' doesn't match overridden member
         public override void Remove(string name)
-#pragma warning restore CS8610
+#pragma warning restore CS8765
```
(FWIW, in situations like this I'm fine with the pragmas; the compiler is correct and we're explicitly violating the contract.  It's the false positives that bother me.)

stephentoub commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/DictionaryDefaultConverter.cs:
```diff
@@ -128,8 +128,9 @@ protected static JsonConverter<TValue> GetValueConverter(ref WriteStack state)
-                        elementConverter.TryRead(ref reader, typeof(TValue), options, ref state, out TValue element);
-                        Add(element, options, ref state);
+                        bool success = elementConverter.TryRead(ref reader, typeof(TValue), options, ref state, out TValue element);
+                        Debug.Assert(success);
```
I'm fine with the assert being added, but it doesn't actually help with any nullable validation, does it?

GrabYourPitchforks replied:
I'm reverting all of the _System.Text.Json_ changes because Layomi's latest PR unifies the nullability attributes across their code base.

stephentoub reviewed (approved):
Thanks for fixing these up.

layomia commented:
The System.Text.Json changes have been addressed in https://github.com/dotnet/runtime/pull/35826, in reaction to `JsonConverter<T>.Read` being annotated with `[return: MaybeNull]`. There are a couple differences though, e.g. in my change, the `out T value` of `JsonConverter<T>.TryRead` is annotated with [`[MaybeNull]`](https://github.com/dotnet/runtime/commit/29a837b8d436dc5925d977a92a77c9de6070b59e#diff-56b52018b267905bd21daf4c7615b6c7L100), rather than [`[MaybeNullWhen(false)]`](https://github.com/dotnet/runtime/commit/29a837b8d436dc5925d977a92a77c9de6070b59e#diff-56b52018b267905bd21daf4c7615b6c7L100).

All things considered, after that PR is merged (which should be very soon), there shouldn't be any changes to System.Text.Json needed here.

danmosemsft commented on src/libraries/System.Net.WebHeaderCollection/ref/System.Net.WebHeaderCollection.cs:
```diff
@@ -97,7 +97,9 @@ public partial class WebHeaderCollection : System.Collections.Specialized.NameVa
         public void Add(System.Net.HttpRequestHeader header, string? value) { }
         public void Add(System.Net.HttpResponseHeader header, string? value) { }
         public void Add(string header) { }
```
All these presumably complicate things next time we run the tool to update the refs.

GrabYourPitchforks commented:
Thanks @layomia for the tip! If your stuff is going in today I'll back my JSON changes out and rebase on top of yours when ready.

layomia commented:
@GrabYourPitchforks https://github.com/dotnet/runtime/pull/35826 is in now (sorry for conflicts ;)

GrabYourPitchforks commented:
Conflicts are fine. I appreciate your efforts @layomia over on the JSON + nullable PR and for driving it through API review! :)

GrabYourPitchforks committed:
Merge remote-tracking branch 'origin/master' into fix_nullability_2

GrabYourPitchforks committed:
PR feedback

GrabYourPitchforks commented:
PR feedback-inspired changes since last iteration:

 * Explicitly pinned the new compiler version
 * Backed out the _System.Text.Json_ changes and merged in Layomi's PR which addresses that project's nullability annotations
 * Removed a stray `Debug.Assert` and cleaned up some comment text around pragmas
